### PR TITLE
feat: purify letrec

### PR DESCRIPTION
**Pull request by magnus-madsen**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze This one I would like to merge. I am less sure about the variable ordering one.

**Comment by magnus-madsen** [comment]
@mlutze I am pretty sure this avoids the `alpha -> alpha + rc` problem... For the current system...

File: main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala [hunk 1]
```diff
@@ -444,6 +444,43 @@ object Unification {
     visit(liftM(fs.head), fs.tail)
   }
 
+  /**
+    * Purifies all effect variables in `tpe` (which should be the type of a letrec-bound function).
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala:448]: But only the ones at the right level?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala:448]: Good point. I thought about it too, but was unsure what to do. It's is probably correct to consider higher levels, but then again there would have been a call to this function for that level. So maybe not necessary. Only doing it for the same level seemed the safest.

**Comment by magnus-madsen** [comment]
@mlutze request approval

**Review by mlutze — APPROVED** [review]

**Pull request merged**